libraries/state_history/include/eosio/state_history/log.hpp

```diff
@@ -68,818 +66,405 @@ static const uint16_t ship_feature_pruned_log = 1;
-            log.skip(-sizeof(pos));
-            fc::raw::unpack(log, pos);
-            log.seek(pos);
-            read_header(header, false);
-            log.seek(pos);
-            EOS_ASSERT(is_ship(header.magic) && is_ship_supported_version(header.magic), chain::plugin_exception, "corrupt ${name}.log (6)", ("name", name));
-
-            index.skip(-sizeof(uint64_t));
-            fc::raw::pack(index, pos);
-            index.skip(-sizeof(uint64_t));
-
-            if (!(remaining % 10000))
-               ilog("${r} blocks remaining, log pos = ${pos}", ("r", remaining)("pos", pos));
-         }
-      }
 
-      index.close();
-      index.open(fc::cfile::create_or_update_rw_mode);
-   }
+      ilog("Regenerate ${name}", ("name", index.display_path()));
+      index.resize(0);
 
-   uint64_t get_pos(uint32_t block_num) {
-      uint64_t pos;
-      index.seek((block_num - _index_begin_block) * sizeof(pos));
-      index.read((char*)&pos, sizeof(pos));
-      return pos;
-   }
+      if(log.size()) {
+         ssize_t next_logpos = log.size() - sizeof(uint64_t);
+         index.resize(expected_index_size);
 
-   void truncate(uint32_t block_num) {
-      log.close();
-      index.close();
+         log_header header = log.unpack_from<decltype(header)>(0);
+         if(is_ship_log_pruned(header.magic))
+            next_logpos -= sizeof(uint32_t);
 
-      auto first_block_num     = catalog.empty() ? _begin_block : catalog.first_block_num();
-      auto new_begin_block_num = catalog.truncate(block_num, log.get_file_path());
+         do {
+            const uint64_t logpos = log.unpack_from<uint64_t>(next_logpos);
+            header = log.unpack_from<decltype(header)>(logpos);
+            EOS_ASSERT(is_ship(header.magic) && is_ship_supported_version(header.magic), chain::plugin_exception, "corrupt ${name}, unknown header magic", ("name", log.display_path()));
 
-      // notice that catalog.truncate() can replace existing log and index files, so we have to
-      // close the files and reopen them again; otherwise we might operate on the obsolete files instead.
+            const uint64_t index_offset_for_bnum = (chain::block_header::num_from_id(header.block_id) - _index_begin_block)*sizeof(uint64_t);
+            if(index.unpack_from<uint64_t>(index_offset_for_bnum) == 0)  //don't overwrite newer blocks for a given blocknum
+               index.pack_to(logpos, index_offset_for_bnum);
 
-      if (new_begin_block_num > 0) {
-         _begin_block = new_begin_block_num;
-         _index_begin_block = new_begin_block_num;
+            next_logpos = logpos - sizeof(uint64_t);
+            if (!(chain::block_header::num_from_id(header.block_id) % 10000))
+               ilog("${r} blocks remaining, log pos = ${pos}", ("r", chain::block_header::num_from_id(header.block_id) - _begin_block)("pos", logpos));
+         } while(chain::block_header::num_from_id(header.block_id) != _begin_block);
       }
```

Comment:
Should we check actual number of indexes generated is the same as `expected_index_size`?

Reply:
Add an ilog saying regeneration is done as we logged when regeneration started at line 384.

Reply:
> Should we check actual number of indexes generated is the same as `expected_index_size`?

Because we work from the end to the front, it looks hard for this to ever not be the case because the very first block we read from the log will have its position placed at the very end of the index immediately satisfying that condition. There could probably be additional robustness checks added to this function, but just checking on the expected index size doesn't really seem to bring much benefit.

> Add an ilog saying regeneration is done as we logged when regeneration started at line 384.

Added as part of 4681ad2